### PR TITLE
BL-13601 receive text message with authentication code

### PR DESCRIPTION
**Pull request by shibuzgeorge**

### CODE REVIEW

File: motr-webapp/webapp/src/main/java/uk/gov/dvsa/motr/web/component/subscription/service/SmsConfirmationManageService.java [hunk 1]
```diff
@@ -0,0 +1,208 @@
...
+                        .setCode(confirmationCode)
+                        .setConfirmationId(confirmationId)
+                        .setAttempts(INITIAL_ATTEMPTS)
+                        .setResendAttempts(INITIAL_RESEND_ATTEMPTS);
+
+                smsConfirmationManageRepository.saveWithResendTimestampUpdate(confirmation);
+                notifyService.sendPhoneNumberConfirmationManageSms(phoneNumber, confirmationCode);
+
+                EventLogger.logEvent(
+                        new SmsConfirmationCreatedEvent().setPhoneNumber(phoneNumber).setConfirmationCode(confirmationCode));
```
Comment (mateuszc-kainos) [inline motr-webapp/webapp/src/main/java/uk/gov/dvsa/motr/web/component/subscription/service/SmsConfirmationManageService.java:81]: I think we should create a separate event for SmsConfirmationManage, e.g. SmsConfirmationManageCreatedEvent

File: motr-webapp/webapp/src/main/java/uk/gov/dvsa/motr/web/component/subscription/service/SmsConfirmationManageService.java [hunk 2]
```diff
@@ -0,0 +1,208 @@
...
+
+                String confirmationCode = generateCode();
+
+                SmsConfirmationManage newSmsConfirmation = copySmsConfirmationManage(smsConfirmation.get())
+                        .setCode(confirmationCode);
+                smsConfirmationManageRepository.save(newSmsConfirmation);
+                notifyService.sendPhoneNumberConfirmationManageSms(phoneNumber, confirmationCode);
+                incrementResendManageAttempts(newSmsConfirmation);
+                EventLogger.logEvent(
+                        new SmsConfirmationCreatedEvent().setPhoneNumber(phoneNumber).setConfirmationCode(confirmationCode));
```
Comment (mateuszc-kainos) [inline motr-webapp/webapp/src/main/java/uk/gov/dvsa/motr/web/component/subscription/service/SmsConfirmationManageService.java:95]: See above comment

File: motr-webapp/webapp/src/main/java/uk/gov/dvsa/motr/web/component/subscription/service/SmsConfirmationManageService.java [hunk 3]
```diff
@@ -0,0 +1,208 @@
...
+            } else {
+
+                motrSession.setSmsConfirmManageResendLimited(true);
+            }
+        } catch (Exception e) {
+
+            EventLogger.logErrorEvent(
+                    new SmsConfrimationCreationFailedEvent().setPhoneNumber(phoneNumber), e);
+            throw e;
```
Comment (mateuszc-kainos) [inline motr-webapp/webapp/src/main/java/uk/gov/dvsa/motr/web/component/subscription/service/SmsConfirmationManageService.java:104]: See above comments

File: motr-webapp/webapp/src/main/java/uk/gov/dvsa/motr/web/component/subscription/service/SmsConfirmationManageService.java [hunk 4]
```diff
@@ -0,0 +1,208 @@
...
+        }
+
+    }
+
+    public Confirmation verifySmsConfirmationManageCode(String phoneNumber, String confirmationId, String confirmationCode)
+            throws InvalidConfirmationIdException {
+
+        SmsConfirmationManage smsConfirmation = smsConfirmationManageRepository.findByPhoneNumber(phoneNumber)
+                .orElseThrow(() -> {
+                    EventLogger.logEvent(new InvalidSmsConfirmationIdUsedEvent().setUsedId(confirmationId).setPhoneNumber(phoneNumber));
```
Comment (mateuszc-kainos) [inline motr-webapp/webapp/src/main/java/uk/gov/dvsa/motr/web/component/subscription/service/SmsConfirmationManageService.java:114]: See above comments

File: motr-webapp/webapp/src/main/java/uk/gov/dvsa/motr/web/component/subscription/service/SmsConfirmationManageService.java [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+import java.util.Optional;
+
+import javax.inject.Inject;
+
+import static uk.gov.dvsa.motr.web.component.subscription.service.ConfirmationCodeGenerator.generateCode;
+
+import static java.time.temporal.ChronoUnit.SECONDS;
+
+
+public class SmsConfirmationManageService {
```
Comment (mateuszc-kainos) [inline motr-webapp/webapp/src/main/java/uk/gov/dvsa/motr/web/component/subscription/service/SmsConfirmationManageService.java:27]: There seems to be a lot of duplication and similarities to SmsConfirmationService - can we somehow use the common code and make it cleaner? Some kind of flag that would tell us which path we are taking - SmsConfirmation or SmsConfirmationManage - that would then influence which repositories, notifytemplates, etc. should be used?

File: motr-webapp/webapp/src/main/java/uk/gov/dvsa/motr/web/resource/CheckYourPhoneResource.java [hunk 6]
```diff
@@ -54,27 +56,163 @@ public Response checkYourPhonePageGet() throws Exception {
...
+
+            modelMap.put(MESSAGE_MODEL_KEY, validator.getMessage());
+            modelMap.put(MESSAGE_AT_FIELD_MODEL_KEY, validator.getMessageAtField());
+            modelMap.put("showInLine", true);
+        }
+
+        updateMap(modelMap);
+        modelMap.put("title_text", "Check your phone");
+
+
```
Comment (mateuszc-kainos) [inline motr-webapp/webapp/src/main/java/uk/gov/dvsa/motr/web/resource/CheckYourPhoneResource.java:108]: Double whitespace

**Review by mateuszc-kainos — CHANGES_REQUESTED** [review]

File: motr-webapp/webapp/src/main/java/uk/gov/dvsa/motr/web/component/subscription/service/SmsConfirmationManageService.java (hunk 5 above)
Comment (shibuzgeorge) [reply motr-webapp/webapp/src/main/java/uk/gov/dvsa/motr/web/component/subscription/service/SmsConfirmationManageService.java:26]: Tried adding flag checks to SmsConfirmationService to make it one class but there is still changes in SmsConfirmationManage that cannot be used for both - such as not needing the confirmation id session and also generating new security code each time they click resend. It didn't make much of a difference, just made it harder to read

**Review by mateuszc-kainos — APPROVED** [review]

**Review by aaronddvsa — APPROVED** [review]

**Pull request merged**